### ISSUE TITLE
[Feature Request] Schedule PV_curtailment later in the day

### ISSUE THREAD

**Issue by purcell-lab** [davidusb-geek/emhass#342]
**Describe the bug**
PV_curtailment is often scheduled early in the day.

**To Reproduce**
Schedule data with a high level of PV_curtailment

**Expected behavior**
It would be better allocation of energy resources to schedule the curtailment as the last event in the day, after battery charging and allocation to all deferrable loads has been completed.

**Screenshots**
![EMHASS-–-Home-Assistant (4)](https://github.com/user-attachments/assets/11ac2c85-7993-43e0-8369-73f8dafe421c)

![image](https://github.com/user-attachments/assets/345f5e9f-053d-406e-94c1-a8c240da5d8d)

**Home Assistant installation type**
 - Home Assistant OS

**Your hardware**
- OS: HA OS
- Architecture: amd64

**EMHASS installation type**
 - Add-on

**Comment by TD944** [comment]
It's definitely something that would be great to have. There seem to be some small factors that influence the distribution as through the day I find the MPC varies between stacking the charge in the morning and then through the day. I'm yet to work out what that is! Even a simple no charging until 100% would be great. 

I wonder too if automatically computing curtailment can fix some infeasible problems. My setup would error on days with high solar generation and low load (e.g. EV away) but was fine when computing curtailment.

In the meantime my battery controller charges the battery if less than maximum and curtailment scheduled.

**Comment by tbrasser** [comment]
Yeah this would be great! I too believe this would result in desirable behaviour/planning.

In my case (cannot export, 4 days horizon), it would mean charging to 100%, then curtailment only later in the horizon, instead of already curtailing straight away skewing the forecast (mainly soc forecast).

**Comment by purcell-lab** [comment]
## Status update — May 2026

**Findings**

- @TD944 and @tbrasser have echoed this from independent angles in April 2026 (cannot-export configurations, MPC re-planning oscillating between morning-stacking and through-day curtailment, soc-forecast skew). The pattern is consistent: the optimiser is allowed to **substitute** curtailment for any other energy sink at any time of day, so it sometimes picks "curtail now" even when "store now, curtail later" gives the same objective value.
- @TD944's secondary observation is interesting and probably useful: enabling automatic curtailment computation also avoided infeasibility on high-solar / low-load days. So curtailment isn't just a scheduling-aesthetics issue — it's also a feasibility safety net.

**Why this happens (best guess)**

`P_PV_curtailment` is a non-negative variable bounded above by available PV, with no explicit time-of-day cost. The objective function only sees the energy balance, so among the (often many) optimal solutions, the solver picks whichever one its branching strategy hits first. This is a classic case of **degenerate optimal solutions** that need a tie-breaker.

**Proposed way forward** — options in order of increasing diff size:

1. **Small lexicographic tie-breaker (cheapest fix).** Add a tiny secondary cost to `P_PV_curtailment[t]` that decreases with `t` over the optimisation horizon — e.g. `epsilon * (T - t) * P_PV_curtailment[t]` where `epsilon` is small enough not to perturb the primary objective. This pushes the solver to delay curtailment whenever it's free to do so, without changing the headline cost.
2. **Hard ordering constraint.** Forbid curtailment in any timestep where the battery isn't full and all deferrable loads with remaining energy budget are off. This is more correct but harder to express cleanly in the LP.
3. **MPC consistency penalty.** Add a small cost on changes between successive plans to stop the day-stacking / through-day flipping that @TD944 describes. Same family of fix as the well-known MPC "control penalty" trick.

Option 1 is a 5-line change and should produce the desired "curtail last" behaviour in the vast majority of cases. Option 3 would need to live in the MPC layer and is the harder change.

Happy to put up a PR for option 1 if @davidusb-geek thinks the approach is reasonable.

**Comment by davidusb-geek** [comment]
Ok, let's go with option 1, this seems to be the easiest option without adding more constraints or setup parameters

**Comment by LesIT1** [comment]
@purcell-lab, this one's been on my radar too. The option 1 approach you sketched, a small time-weighted tie-break that pushes curtailment to the latest feasible timesteps, is exactly right, and David gave it the nod back in May, so it seems a shame to let it sit.

Are you still planning to pick it up? Full credit to you for the diagnosis and the design either way. If you'd rather not, I'm happy to carry it forward with credit to you. I run with compute_curtailment on every day, so I can test it against a real curtailing setup before it goes up. Just didn't want to step on your toes if it's already in progress.

**Comment by purcell-lab** [comment]
> Are you still planning to pick it up? Full credit to you for the diagnosis and the design either way. If you'd rather not, I'm happy to carry it forward with credit to you. I run with compute_curtailment on every day, so I can test it against a real curtailing setup before it goes up. Just didn't want to step on your toes if it's already in progress.

Thanks for the ping on this, I have been following your massive push into EMHASS on multiple issues and new functionalities. You have some real momentum  which is fantastic for the project, if not a little difficult to keep up with all the changes coming down the pipeline.

Happy for you to proceed on this issue asking with your other fixes. Great work.
